### DISCUSSION TITLE
Locking & Transactions

### DISCUSSION THREAD

**Discussion by commi235** [flowsforapex/apex-flowsforapex#210] [Ideas]
### General Rules for Transaction Boundaries

- Each Subflow is a separate transaction
- Forking objects will do the forks, mark itself completed and commit.
- Forks start up as "zombies" and will get active only after parent (forking object) is done.
- Each fork gets its own transaction (as they are subflows)

**Comment by rallen2010** [comment]
- [x] Create new Github Branch feature/transaction-safety
#### Subflow Status
- [x] Add a new subflow status of 'created' which is used before it is 'running'
- [x] Modify BPMN Viewer Plugin to Add 2 more Content sets.  Can we assign a colour to the content set in the plugin without hitting the 15 attribute limote?  Maybe just for the customer specified one...(-> css)?
- [x] Modify flow_instance_details_vw to add 'Stalled Subflows' or 'error subflows' as a content set, coloured red

#### Lock Subflows before Use each time (select ...for update)
- [x] when flow_complete_step starts on a subflow
- [x] after creation but before stepping off on new subflows after a splitting inclusive, parallel & event-based gateway
- [x] before a merging gateway parallel or inclusive gateway, lock the merging subflows and the 'split' parent subflow
- [ ] on an event-based gateway, in the event handler handling the returned event, lock the parent & all child subflows
- [x] after process_start, lock the subflow
- [ ] after non-interrupting boundary events (not timers) create a new subflow (error, escalation create a new sbfl when they are triggered) , lock it before stepping off. 
- [x] non-interrupting timer boundary events - timer set - the subflow is created when its parent object (task, subProc) becomes the current object.  The boundary event subflow is created as part of the pre-tasks for it's parent object & are part of that transaction.  So no need for explicit locking at this point.
- [x]  non-interrupting timer boundary events - timer firing - sbfl will need to be locked by the firing timer.
- [x]  non-interrupting timer boundary events - timer unset - performed as part of the post-tasks for its parent object.
- [ ] before clearing up / terminating subflows as a result of a sub process terminating (interrupting boundary condition, terminating endEvent). Clearing up levels of subProceses is done recursively, so modify procedure to know whether it was called externally or recursively.  Lock all in levels on way down, do deletes, & commit when back to the top.
- [x] before reset process
- [x] before delete process



#### Commits:
Transaction Boundaries inside Flows for APEX should be on each step forward of the model.

Process Operations:
- [x] Create process & then commit
- [x] Start Process - Lock Process, Create Subflow in 'running' status, execute pre expressions, call step_complete, which commits after post phase steps.
- [x] Reset Process - Lock, reset, commit.
- [x] Delete Process - Lock, reset, commit.

For all Process steps:
> - [x] Inside Flow_complete_step there is a boundary between post- tasks on the  current  step, and pre- tasks for the next step.  That should be the commit point for each model step, and commits the old current step.  
> - [x] At the end of flow_complete_step, after processing the step & before we return to the app, commit.  For simple tasks, this will just commit the step pre-tasks.

Additionally, for specific object types there are additional transactions as part of object processing :
- [x] Splitting Gateways (Parallel, Inclusive, Event-Based) should commit after creating their forked subflows and marking themselves as Split -- and before stepping into each of the forked subflows as separate transactions.
- [x] Merging Gateways (Parallel, Inclusive) should commit after completing the merge and marking the 'Split' parent transaction as 'running' again.  The parent steps forward as a new transaction.
- [x] non-Interrupting Boundary Events should commit at step complete (like any other step)
- [ ] interrupting Boundary Events should commit after moving the flow into its new path (which is like any other step)
- [ ] After the pre-section & of a scripttask (so before passing to the script)
- [ ] After the script has run in a scriptTask 
- [ ] once a timer is scheduled
- [x] A link event transaction spans the throw and catch event of a link as a single transaction.

Other Functions:
- [x] reservations & releases of reservations.  Note releases will need be separate tx when called externally, but part of the larger complete_step tx when called internally.
### Open Issues / questions
- process variables?

**Comment by rwest-tm** [comment]
The flow_timers_pkg.step_timers procedure will need to be restructured to use select for update as well. Suggested code for the step timers is shown below. It expects the flow_engine.flow_handle_event to commit changes and raise an exception in the subflow is not found.
```
  procedure step_timers
  as
    e_resource_timeout      exception;
    pragma                  exception_init(e_resource_timeout, -30006);

    l_timers                flow_timers%rowtype;
    l_run_time              flow_timers.timr_last_run%type := systimestamp;
  begin
    loop -- until no records found
        -- could add a functional index on flow_timers to improve performance of this query
        -- eg. create index flow_timr_n1 on flow_timers (
        --          case when timr_status in ( 'C', 'A' ) then coalesce( timr_last_run, timr_created_on ) end);
        select * into l_timers
        from flow_timers
        where rowid in (
            select max(rowid) keep (dense_rank first order by coalesce( timr_last_run, timr_created_on )) trowid
            from flow_timers
            where case when timr_status in ( c_created, c_active ) then coalesce( timr_last_run, timr_created_on ) end < l_run_time 
            and case timr_type
                    when flow_constants_pkg.gc_timer_type_cycle then
                        timr_start_on
                            + ( timr_interval_ym * coalesce( timr_run_count, 1 ) )
                            + ( timr_interval_ds * coalesce( timr_run_count, 1 ) )
                    else timr_start_on 
                end <= l_run_time
            )
        for update wait 5
        ;
        
        case l_timers.timr_type when flow_constants_pkg.gc_timer_type_cycle then 
            l_new_status := case when l_timers.timr_run_count + 1 = l_timers.timr_repeat_times then c_ended else c_active end;
        else 
            l_new_status := c_ended;
        end case;

        update flow_timers
        set timr_last_run = systimestamp
            , timr_run_count = timr_run_count + 1
            , timr_status = l_new_status
        where timr_id = l_timers.timr_id
        ;
                
        begin
        -- ideally the flow_engine should lock the subflow and this procedure should handle the resource 
        -- timeout, deadlock and not found exceptions. This would happen if the subflow is locked waiting 
        -- to delete the timer through a cascade delete.
        flow_engine.flow_handle_event
            (
              p_process_id => l_timers.timr_prcs_id
            , p_subflow_id => l_timers.timr_sbfl_id
            );
        exception 
            -- Some exception happened during processing the timer
            -- We trap it here and mark respective timer as broken.
            when others then
            apex_debug.error
            (
                p_message => 'Timer with ID %s did not work. Check logs.'
            , p0        => l_timers.timr_id
            );
            update flow_timers
                set timr_status = c_broken
            where timr_id = l_timers.timr_id
            ;
        end;
        commit;
    end loop;
    exception 
        when no_data_found then return;
        when e_resource_timeout then
            -- record requiring update is locked by another process, could put some logging in here
            rollback;
            return;

  end step_timers;
```

**Comment by rwest-tm** [comment]
To ensure transactions do not deadlock it is important to lock records in a consistent order. So where a parent record needs to be updated and a child record inserted, updated or deleted then the parent record should be locked with an update or a select for update before the child records are inserted, updated or deleted. This ensures that the state of the parent and child records are consistent and deadlocks are avoided.

The hierarchy of flow records is:
flow_processes -> main flow_subflows (sbfl_sbfl_id = null) -> child flow_subflows (sbfl_sbfl_id is not null) -> grandchild flow_subflows etc.
There are other subrecords in the flow_subflow_log, flow_process_variables and flow_timers that will need to be considered as well.

Currently the locking in the flow_start and flow_reset APIs could allow a subflow record to be inserted but the process left in the created state. This could happen if the flow_reset was started before the flow_start transaction had been commited. The flow_reset update flow_processes statement would wait until the flow_start transaction was commited setting the process status back to created but leaving the subflow record inserted by the flow_start in place. The flow_reset API should update the flow_processes record to created then delete the flow_subflows records. The flow_start API should select for update on the flow_process then if the status is created insert the subflow, then update the process to running.

This locking strategy should be applied consistently in each transaction where subflow records are being inserted and deleted. For this reason completing a subprocess will need to be in a separate transaction.

**Reply by rallen2010** [reply]
I marked some items 'done' above but haven't committed or sync'd them back into the GitHub server yet as I'm not quite done testing ...so you're not seeing the full start and reset processes yet... 
Start_process updates the process and creates the subflow in one transaction.  
Then calls complete_step on the subflow - which locks it & does the -post step actions and commits
and then locks the subflow again before stepping into the next task....

**Comment by rwest-tm** [comment]
Another consideration is the exception handling that needs to be put in place to handle any exceptions from a select for update statement. Without a wait clause on the select for update the statement will wait for ever if the record is already locked by another transaction. This is probably not the desired response. A nowait would result in unwanted exceptions when a record is temporarily locked by another transaction, so a wait for a few seconds is probably the best option. If select for update statement times out when getting the lock then the transaction should be rolled back and an exception raised. 

The select for update will also need to handle a no_data_found exception. In most cases this can be ignored because the flow that needed to be updated has been deleted.

The select for update may also need to raise an exception if the record is in an unexcepted state, for example, excepted a status of running but the locked record is in a different state.
